### PR TITLE
Add CategoryCollection::getDescendantIDs

### PR DESCRIPTION
**Pull request by cchabilall83**
Required for https://github.com/vanilla/internal/pull/2612

- Add `getDescendantIDs` method to categoryCollection class.
- Modifies `expandCategories` to look for `ParentRecordID` if `CategoryID` not available.

### CODE REVIEW

File: applications/vanilla/library/class.categorycollection.php [hunk 1]
```diff
@@ -461,6 +461,42 @@ public function getTree($parentID = -1, $options = []) {
...
+        $parentIDs = [$parentIDs] ?: [-1];
+        $defaultOptions = [
+            'maxDepth' => 3,
+            'permission' => 'PermsDiscussionsView'
+        ];
+        $options = array_change_key_case($options) ?: [];
+        $options = $options + $defaultOptions ;
+
+        for ($i = 0; $i < $options['maxDepth']; $i++) {
+            $childCategories = $this->getChildrenByParents($parentIDs, $options['permission']);
```
Comment (charrondev) [inline applications/vanilla/library/class.categorycollection.php:482]: This makes a request to the DB every time it's called.

File: applications/vanilla/library/class.categorycollection.php [hunk 2]
```diff
@@ -461,6 +461,42 @@ public function getTree($parentID = -1, $options = []) {
...
+    /**
+     * Get the id's of a category's descendants.
+     *
+     * @param int $parentIDs
+     * @param array $options
+     * @return array
+     */
+    public function getDescendantIDs($parentIDs = -1, $options = []) {
+        $ids = [];
+        $parentIDs = [$parentIDs] ?: [-1];
```
Comment (charrondev) [inline applications/vanilla/library/class.categorycollection.php:473]: This will never hit it's fallback condition.

Comment (charrondev) [reply applications/vanilla/library/class.categorycollection.php:473]: Shouldn't it be called `parentID`?

File: applications/vanilla/library/class.categorycollection.php [hunk 3]
```diff
@@ -461,6 +461,42 @@ public function getTree($parentID = -1, $options = []) {
...
     }
 
+    /**
+     * Get the id's of a category's descendants.
+     *
+     * @param int $parentIDs
+     * @param array $options
+     * @return array
+     */
+    public function getDescendantIDs($parentIDs = -1, $options = []) {
```
Comment (charrondev) [inline applications/vanilla/library/class.categorycollection.php:471]:
```suggestion
    public function getDescendantIDs(int $parentID = -1, array $options = []) {
```

File: applications/vanilla/library/class.categorycollection.php [hunk 4]
```diff
@@ -461,6 +461,42 @@ public function getTree($parentID = -1, $options = []) {
...
+        $ids = [];
+        $parentIDs = [$parentIDs] ?: [-1];
+        $defaultOptions = [
+            'maxDepth' => 3,
+            'permission' => 'PermsDiscussionsView'
+        ];
+        $options = array_change_key_case($options) ?: [];
```
Comment (charrondev) [inline applications/vanilla/library/class.categorycollection.php:478]: What's this for? And again I don't think we'll ever had the fallback of the ternary.

File: applications/vanilla/library/class.categorycollection.php [hunk 5]
```diff
@@ -461,6 +461,36 @@ public function getTree($parentID = -1, $options = []) {
...
     }
 
+    /**
+     * Get the id's of a category's descendants.
+     *
+     * @param int $parentID
+     * @param array $options
+     * @return array
+     */
+    public function getDescendantIDs($parentID = -1, $options = []) {
```
[the remark this answers is not in the document]
Comment (cchabilall83) [reply applications/vanilla/library/class.categorycollection.php]: done

File: applications/vanilla/library/class.categorycollection.php (hunk 1 above)
Comment (cchabilall83) [reply applications/vanilla/library/class.categorycollection.php]: added caching

File: applications/vanilla/library/class.categorycollection.php (hunk 2 above)
Comment (cchabilall83) [reply applications/vanilla/library/class.categorycollection.php]: done

File: applications/vanilla/library/class.categorycollection.php (hunk 4 above)
Comment (cchabilall83) [reply applications/vanilla/library/class.categorycollection.php]: removed it

**Comment by charrondev** [comment]
We're going to hold on tests on this new methods. I'm likely to come through and do some refactoring on this model at the end of this sprint, and that will come with an extended test suite.

**Pull request merged**